apps/alert_processor/lib/lock.ex

```diff
@@ -0,0 +1,71 @@
+defmodule AlertProcessor.Lock do
+  @moduledoc "Provides a database-backed exclusive lock."
+  alias AlertProcessor.Repo
+
+  # Arbitrary integer; should be unique among all locks used with this database
+  @key 0
+
+  @doc """
+  Tries to acquire an exclusive advisory lock using the `AlertProcessor.Repo` database. The given
+  function will be called with `:ok` if the lock was acquired, or `:error` if the lock is already
+  in use (i.e. a function that was called with `:ok` is currently executing elsewhere).
+
+  Returns whatever the passed function returns.
+
+  Note: The function is not executed in a transaction, allowing multiple transactions to be
+  performed "within" a lock. The lack of a `Repo.checkout` in Ecto 2 prevents us from using a
+  session-level lock to do this, so instead we spawn a process that acquires a transaction-level
+  lock and holds the transaction open until the passed function is done executing.
+  """
+  @spec acquire((:ok | :error -> any)) :: any
+  def acquire(func) do
+    parent = self()
+
+    locker =
+      spawn_link(fn ->
+        Process.flag(:trap_exit, true)
+
+        Repo.transaction(
+          fn ->
+            case Repo.query!("SELECT pg_try_advisory_xact_lock($1)", [@key]) do
+              %{rows: [[true]]} ->
+                send(parent, :acquired)
+
+                receive do
+                  :release -> nil
+                  {:EXIT, _, _} -> nil
```

Comment:
> Does this handle any case where the process holding a lock dies or is killed?

In general yes, though how it happens depends on what you mean by "the process holding a lock".

1. If any code in _this_ process (the "locker") raises an exception somehow, `Repo.transaction` should guarantee the transaction is rolled back, which releases the advisory lock.
2. If this process dies or is killed, the `trap_exit` flag above means it will get an `:EXIT` message here, releasing it from the `receive` and closing the transaction, which again releases the advisory lock.
3. If the process calling `acquire` raises an exception inside the function it passed to `acquire`, the `try ... after` below will notify the "locker", releasing it from the `receive` and closing the transaction as in (2).
4. If the process calling `acquire` dies or is killed, the "locker" will also be terminated due to it having been started with `spawn_link`, and then (2) applies.

The test you linked is specifically testing (3). There's currently no explicit test for (4), though [this test](https://github.com/mbta/alerts_concierge/pull/1145/files#diff-fbfbcd8ddf827e97dff2425505302219eee6815e764ab368378d07e362c67a2fR26) actually tests it indirectly (the test process getting torn down before the `sleep` has finished creates this situation). I think adding a test for it would be good, anyway.

Reply:
👍 thanks, that's a helpful description.